README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

Package for using custom sidebar in flutter as an parent root widget.

## Features

- Custom Sidebar with animation
- Custom Appbar
- Parent widget to cover it all

## Screenshots
![](https://github.com/UsamaYousaf/UsamaYousafgithub.io/blob/main/assets/assets/photos/samplegif.gif?raw=true)


## Getting started
pub.dev
## Usage
Create the List of MenuItems and add the MenuItem which you wish to add in your drawer using this widget

```dart
 List<CustomMenuItem> menuItemsList=[
     CustomMenuItem(
                        callback: () {
                        //Callback function to route to page on Click
                        },
                        title: 'Home',
                        leadingIcon: Icons.home,
                          iconSize: 22,
       titleSize: 16,


                      ),
    CustomMenuItem(
                        callback: () {
                        //Callback function to route to page on Click
                        },
                        title: 'History',
                           leadingIcon: Icons.history,
                                 iconSize: 22,
      titleSize: 16,
                       
                      ),
                      ];

                     

```

Add the List of menuItems in CustomDrawer widget with other customised details .
see example for further information

```dart
Scaffold(
      // appBar: AppBar(

      //   title: Text(widget.title),
      // ),
      body:CustomDrawer(homeWidget: Center(child:Text('Your Home Widget'),),
      menuItemsList: menuItemsList, appBarActions: [], appBarTitle: Text('Your Home Widget') , menuIcon: const Icon(Icons.menu),
      )
      
      );
```


## Additional information
For contribution and issues: https://github.com/UsamaYousaf/custom_sidemenu


CHANGELOG.md
## [0.0.1] -2022-11-11

* Initial release with basic ui

## [0.0.2] -2022-11-11

* Initial release with readme

## [0.0.3] -2022-11-11

* Initial release with example

test/custom_sidemenu_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:custom_sidemenu/custom_sidemenu.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//   });
// }


lib/custom_sidemenu.dart
library custom_sidemenu;

import 'package:flutter/material.dart';

class CustomDrawer extends StatefulWidget {
  CustomDrawer(
      {Key? key,
      required this.menuItemsList,
      required this.homeWidget,
      required this.appBarTitle,
      required this.appBarActions,
      required this.menuIcon,
      this.elevation,
      this.padding,
      this.background})
      : super(key: key);

  /// menu item list
  List<CustomMenuItem> menuItemsList;

  /// title of appbar
  Widget appBarTitle;

  /// Actions of appbar
  List<Widget> appBarActions;

  /// home page widget,
  Widget homeWidget, menuIcon;

  /// elevation
  double? elevation;

  /// padding for list items
  EdgeInsets? padding;

  /// background color

  Color? background;

  @override
  CustomDrawerState createState() => CustomDrawerState();
}

class CustomDrawerState extends State<CustomDrawer>
    with SingleTickerProviderStateMixin {
  late AnimationController animationController;

  bool visible = false;

  @override
  void initState() {
    super.initState();
    animationController = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 250));
  }

  @override
  Widget build(BuildContext context) {
    Widget _buildDrawer() => Drawer(
          elevation: widget.elevation,
          backgroundColor:
              widget.background ?? const Color.fromARGB(255, 28, 92, 146),
          child: SingleChildScrollView(
            physics: const BouncingScrollPhysics(),
            child: Container(
                height: MediaQuery.of(context).size.height,
                padding:
                    widget.padding ?? const EdgeInsets.fromLTRB(30, 90, 90, 0),
                child: ListView.builder(
                  itemBuilder: (context, index) => widget.menuItemsList[index],
                  itemCount: widget.menuItemsList.length,
                )),
          ),
        );

    Widget _buildHome(Widget homeWidget) => Container(
        decoration: BoxDecoration(boxShadow: [
          BoxShadow(
            color: Colors.black45.withOpacity(0.25),
            spreadRadius: 10,
            blurRadius: 7,
            offset: const Offset(0, 3), // changes position of shadow
          ),
        ], borderRadius: BorderRadius.circular(25)),
        child: HomeWidget(
          title: widget.appBarTitle,
          appBarActions: widget.appBarActions,
          menuIcon: widget.menuIcon,
          callBack: toggleAnimation,
          child: homeWidget,
        ));
    return AnimatedBuilder(
        animation: animationController,
        builder: (context, _) {
          double slide = 200.0 * animationController.value;
          double scale = 1 - (animationController.value * 0.15);
          return Stack(
            children: [
              _buildDrawer(),
              Transform(
                  transform: Matrix4.identity()
                    ..translate(slide)
                    ..scale(scale),
                  alignment: Alignment.center,
                  child: GestureDetector(
                    child: _buildHome(widget.homeWidget),
                    onTap: () {
                      //  animationController.reverse();
                    },
                    onPanUpdate: (details) {
                      // if (details.delta.dx < 0) {
                      //   animationController.reverse();
                      // }
                    },
                  )),
            ],
          );
        });
  }

  void toggleAnimation() {
    animationController.isDismissed
        ? animationController.forward()
        : animationController.reverse();
  }
}

class CustomMenuItem extends StatelessWidget {
  const CustomMenuItem({
    Key? key,
    required this.title,
    required this.callback,
    this.leadingIcon,
    this.iconColor,
    this.iconSize,
    this.titleSize,
    this.titleColor,
    this.overflow,
    this.trailingIcon,
    this.contentPadding,
  }) : super(key: key);

  /// menu item title

  final String title;

  /// menu item callback

  final VoidCallback callback;

  /// leading icon

  final IconData? leadingIcon;

  /// tailing icon

  final IconData? trailingIcon;

  /// color of icon default icon color is white
  final Color? iconColor;

  /// icon size default size is 16
  final double? iconSize;

  /// size of title
  final double? titleSize;

  /// color of tittle
  final Color? titleColor;

  /// title overflow default Clip used

  final TextOverflow? overflow;

  /// content padding if  null,EdgeInsets.symmetric(horizontal:16) used
  final EdgeInsetsGeometry? contentPadding;

  @override
  Widget build(BuildContext context) {
    return ListTile(
      onTap: callback,
      trailing: Icon(
        trailingIcon,
        color: iconColor ?? Colors.white,
        size: iconSize ?? 20,
      ),
      leading: Icon(
        leadingIcon,
        color: iconColor ?? Colors.white,
        size: iconSize ?? 20,
      ),
      title: Text(
        title,
        overflow: overflow ?? TextOverflow.clip,
        style: Theme.of(context).textTheme.headline2?.copyWith(
            color: titleColor ?? Colors.white, fontSize: titleSize ?? 10),
      ),
      contentPadding: contentPadding,
    );
  }
}

class HomeWidget extends StatelessWidget {
  const HomeWidget(
      {Key? key,
      required this.title,
      required this.child,
      required this.appBarActions,
      required this.menuIcon,
      required this.callBack})
      : super(key: key);
  final Widget title;

  final Widget child;
  final List<Widget> appBarActions;
  final Widget menuIcon;
  final VoidCallback callBack;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBarWidget(
        title: title,
        menuIcon: InkWell(
          onTap: callBack,
          child: menuIcon,
        ),
        actions: appBarActions,
      ),
      body: child,
    );
  }
}

class AppBarWidget extends StatelessWidget implements PreferredSizeWidget {
  AppBarWidget({
    Key? key,
    required this.title,
    required this.menuIcon,
    this.orientation,
    required this.actions, this.background,

    // ignore: non_constant_identifier_names
  }) : super(key: key);

  final Widget title;
  final Widget menuIcon;
  final List<Widget> actions;
  final Color ? background;

  Orientation? orientation;

  @override
  Widget build(BuildContext context) {
    return AppBar(
        leading: menuIcon,
        title: title,
        actions: actions,
        backgroundColor:background?? const Color.fromARGB(255, 28, 92, 146));
  }

  @override
  Size get preferredSize => const Size.fromHeight(50);
}


example/lib/main.dart
import 'package:custom_sidemenu/custom_sidemenu.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
       
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  
  List<CustomMenuItem> menuItemsList=[
     CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'Home',
                        leadingIcon: Icons.home,
                          iconSize: 22,
       titleSize: 16,


                      ),


    CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'History',
                           leadingIcon: Icons.history,
                                 iconSize: 22,
      titleSize: 16,
                       
                      ),
    CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'Setting',
                          leadingIcon: Icons.settings,
                              iconSize: 22,
      titleSize: 16,
                      ),
    CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'profile',
                           leadingIcon: Icons.account_circle_outlined,
                            iconSize: 22,
      titleSize: 16,
                      ),
    CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'Help',
                         leadingIcon: Icons.help,
                            iconSize: 22,
      titleSize: 16,
                       
                      ),
    CustomMenuItem(
                        callback: () {
                        
                        },
                        title: 'Log Out',
                       leadingIcon: Icons.logout,
                            iconSize: 22,
                        titleSize: 16,
                       
                      ),
  ];

  @override
  Widget build(BuildContext context) {
  
    return Scaffold(
      // appBar: AppBar(

      //   title: Text(widget.title),
      // ),
      body:CustomDrawer(homeWidget: Center(child:Text('Your Home Widget'),),
      menuItemsList: menuItemsList, appBarActions: [], appBarTitle: Text('Your Home Widget') , menuIcon: const Icon(Icons.menu),
      )
      
      );
 
  }
}
